### DISCUSSION TITLE
Prevent arbitrary _id injection on array of subdocuments when updating parent

### DISCUSSION THREAD

**Discussion by gterras** [Automattic/mongoose#12767] [Q&A] [answered]
With a model containing a prop which is an array of subdocuments: 

```ts
class Item {
   subItems: SubItems[]
}
```

and an operation updating the parent: 
```ts
await this.itemModel
    .findOneAndUpdate({ _id: payload.id }, payload, { new: true })
    .exec()
```

Is there a way (in a single query) to prevent injecting a subddocument with an arbitrary (non mongo-defined) `_id` ? 

For example for an existing `Item`: 
```ts
{
 _id: "parent_id",
 subItems: [{_id: 'xxx', label: 'hey'}]
}
```

 and a payload like so is sent for update: 
```ts
const payload = {
  _id: "parent_id"
  subItems: [
    {_id: 'xxx', label: 'new label'}, 
    // here yyy is a valid but non-existing ObjectID
    {_id: 'yyy', label: 'another label'}
  ]
}
```
Then the second subitem will now have `yyy` as id which could be an arbitrary user-defined value (which I don't want).

I could add a `findOne` query beforehand and check the result against `payload` to "manually" to remove  subdocuments unknown `_id` props but is there a more optimized way to do so in a single query ? Thanks

**Comment by ViniciusMarchi** [comment]
I don't know but maybe you can use `$set` operator for this purpose. This operator sets the value of a field in the document that matches the query condition. In this example, maybe we can use this operator to update the subItems field replacing the existing subItems, but handling only documents that have a know `_id`.

I don't know if works well, but you can try something like this:

using the payload example that you show 
```javascript
const payload = {
  _id: 'parent_id',
  subItems: [
    { _id: 'xxx', label: 'new label' },
    { _id: 'yyy', label: 'another label' }
  ]
};
```

run that query:

```javascript
await this.itemModel
  .findOneAndUpdate({ _id: payload.id }, { $set: { subItems: payload.subItems } }, { new: true })
  .exec()
```

**Reply by vkarpov15** [reply]
"I could add a findOne query beforehand and check the result against payload to "manually" to remove subdocuments unknown _id props but is there a more optimized way to do so in a single query" <-- Unfortunately there's no way to do that in a single query. It's up to you to enforce referential integrity in this case.

**Reply by gterras** [reply]
> Unfortunately there's no way to do that in a single query. It's up to you to enforce referential integrity in this case.

Thanks for letting me know.

**Comment by gterras** [comment, marked as answer]
Answered above
